chapter_preliminaries/calculus.ipynb

```diff
@@ -0,0 +1,401 @@
+    "beyond the exact set of data examples used to train them.\n",
+    "\n",
+    "To help you understand\n",
+    "optimization problems and methods in later chapters,\n",
+    "here we give a very brief primer on differential calculus\n",
+    "that is commonly used in deep learning.\n",
+    "\n",
+    "\n",
+    "## Derivatives and Differentiation\n",
+    "\n",
+    "We begin by addressing the calculation of derivatives,\n",
+    "a crucial step in nearly all deep learning optimization algorithms.\n",
+    "In deep learning, we typically choose loss functions\n",
+    "that are differentiable with respect to our model's parameters.\n",
+    "Put simply, this means that for each parameter,\n",
+    "we can determine how rapidly the loss would increase or decrease,\n",
+    "were we to *increase* or *decrease* that parameter\n",
+    "by an infinitesimally small amount.\n",
+    "\n",
+    "Suppose that we have a function $f: \\mathbb{R} \\rightarrow \\mathbb{R}$,\n",
+    "whose input and output are both scalars.\n",
+    "The *derivative* of $f$ is defined as\n",
+    "\n",
+    "$$f'(x) = \\lim_{h \\rightarrow 0} \\frac{f(x+h) - f(x)}{h},$$\n",
+    ":eqlabel:`eq_derivative`\n",
+    "\n",
+    "if this limit exists.\n",
+    "If $f'(a)$ exists,\n",
+    "$f$ is said to be *differentiable* at $a$.\n",
+    "If $f$ is differentiable at every number of an interval,\n",
+    "then this function is differentiable on this interval.\n",
+    "We can interpret the derivative $f'(x)$ in :eqref:`eq_derivative`\n",
+    "as the *instantaneous* rate of change of $f(x)$\n",
+    "with respect to $x$.\n",
+    "The so-called instantaneous rate of change is based on\n",
+    "the variation $h$ in $x$, which approaches $0$.\n",
+    "\n",
+    "To illustrate derivatives,\n",
+    "let us experiment with an example.\n",
+    "Define $u = f(x) = 3x^2-4x$.\n",
+    "\n",
+    "*Note: We will be using Double in this section to avoid incorrect results since Double provides more decimal precision. Generally though, we would use Float as deep learning frameworks by default use Fault.*"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "// %mavenRepo snapshots https://oss.sonatype.org/content/repositories/snapshots/\n",
+    "\n",
+    "%maven ai.djl:api:0.9.0\n",
+    "%maven org.slf4j:slf4j-api:1.7.26\n",
+    "%maven org.slf4j:slf4j-simple:1.7.26\n",
+    "\n",
+    "%maven ai.djl.mxnet:mxnet-engine:0.9.0\n",
+    "%maven ai.djl.mxnet:mxnet-native-auto:1.7.0-backport"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%load ../utils/plot-utils\n",
+    "%load ../utils/Functions.java"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import ai.djl.ndarray.*;\n",
```

Comment:
nit: could you move all the imports to the beginning. I saw other chapters are doing so. Let's follow it

Reply:
Alright yep I'll do that and also verify if the other notebooks are the same.